Worktree support

RickyDaMa opened the pull request:
Fixes #169

Adds git worktree support to font-v, as well as improving the robustness of `.git/` folder detection

<details><summary>Old, out-dated description</summary>
I'm not saying it's a good implementation, but it's something that works to start this conversation

Unhelpfully, the [documentation link](https://gitpython.readthedocs.io/en/stable/reference.html#git.repo.fun.find_worktree_git_dir) @madig recommended was just a page with headings and no actual docs for me to do this properly. I'd be all for using the library's support for this if I could actually see how to :/
</details>

RickyDaMa committed:
get_git_root_path: support worktrees

RickyDaMa committed:
Test worktree support

alerque commented on lib/fontv/utilities.py:
```diff
@@ -44,8 +44,18 @@ def get_git_root_path(filepath):
     # search up to five directories above for the git repo root
     for _ in range(6):
-        if dir_exists(os.path.join(gitroot_path, ".git")):
+        dot_git = os.path.join(gitroot_path, ".git")
```
Before you assume that `.git` exists at the root of a repository, first check `$GIT_DIR` from the environment. The defaults are obviously in common use but exceptions are not so uncommon that it is smart to just ignore them. Unfortunately anywhere you are running `font-v` might not actually by inside a shell environment where the Git dir has been setup for the benefit of other tools, but it's better than nothing.

RickyDaMa replied:
If I'm not mistaken, this is more of a gripe with the original implementation, rather than the logic I added for worktree support?

I can add a check for `$GIT_DIR` as a fast path before the `for` loop if you'd like? I can validate it with `is_git_dir()`

alerque replied:
Yes, just something I happened to notice when reviewing this.

RickyDaMa replied:
Added some logic for this and a corresponding test

RickyDaMa commented:
Okay, this implementation I'm actually happy with

I used the [Wayback Machine](https://web.archive.org/web/20240205010658/https://gitpython.readthedocs.io/en/stable/reference.html#module-git.repo.fun) & [GitHub](https://github.com/gitpython-developers/GitPython/blob/64ec0b1f86ada5565e30bc21b4ad189c6c4df49e/git/repo/fun.py) to get the actual documentation / source code for GitPython

I changed the test for "is `.git` a folder?" to use GitPython's provided `is_git_dir()` so that's more robust

I switched my manual reading/parsing of the `.git` file for worktrees with GitPython's `find_worktree_git_dir` (which has pretty similar logic to my original more naive implementation lol)

I also updated the test to use a fixture, as otherwise if `get_git_root_path` raised then the worktree wouldn't get cleaned up; now it does

As an aside, not sure if the pipeline failure is my fault and/or how to fix it

RickyDaMa committed:
get_git_root_path: check $GIT_DIR first

chrissimpkins commented:
Thank you very much Ricky. And thanks for the review Caleb.  Let me fix the CI and I'll have a look at this PR this week.  I really appreciate it!
